File: programming/Puzzles/usaco/gift1.java
/*
ID: bill.pi1
LANG: JAVA
TASK: gift1
*/



import java.util.*;
import java.io.*;




public class gift1{


    public static int[] toIntArray(String[] str){
        int[] arr = new int[str.length];
        for(int i = 0; i < str.length; i++){
            arr[i] = (Integer.parseInt(str[i]));
        }
        return arr;
   
    }
    

    //Store info in Block class

    static class Block {
        String giver = "fubar";
        int mn = 0;
        List<String> takers = new ArrayList<String>();
    
        
        public Block(String giver, int mn, List<String> takers) {
            this.giver = giver;
            this.mn = mn;
            this.takers = takers;
        }
    }
    


    //read file into line
    public static List<String> stream(String name) {
        try{
            List<String> lines = new ArrayList<String>();
            Scanner sc = new Scanner(new FileInputStream(name));
            while(sc.hasNextLine()) {
                lines.add(sc.nextLine());
            }

            sc.close();
            return lines;

        }catch(IOException e){
            System.out.println("uh oh");
            return null;
        } 
    }


    public static void blockerizer(List<String> input, List<Block> output){

        if(input.size() > 0) {
            
            String giver = input.get(0);
            int[] data = toIntArray(input.get(1).split(" ")); 
            List<String> takers = input.subList(2, 2+data[1]);
            output.add(new Block(giver, data[0], takers));
            blockerizer(input.subList(2+data[1], input.size()) , output);

        }

    }

    public static void cashier(HashMap<String, Integer> bank, List<Block> blocks){
        for(Block block : blocks){
            
            if(block.mn != 0){
                int a = (int)(block.mn / block.takers.size());
                int b = block.mn - (a*block.takers.size());

                for(String t : block.takers){
                    bank.put(t, bank.get(t)+a);
                }
                bank.put(block.giver, bank.get(block.giver) + b - block.mn);
            }
            
            
        }
    }
    
    public static List<String> banker(String file){

        HashMap<String, Integer> bank = new HashMap<String, Integer>();
        List<String> input = stream(file);
        List<Block> blocks = new ArrayList<Block>();
        List<String> people = input.subList(1, Integer.parseInt(input.get(0)) + 1);
        
        for(String p : people){
            bank.put(p, 0);
        }
        blockerizer(input.subList(people.size()+1, input.size()), blocks);
        cashier(bank, blocks);

        for(int i = 0; i < people.size(); i++){
            String p = people.get(i);
            people.remove(i);
            people.add(i, p + " " + bank.get(p));
        }

    return people;

    }


    public static void main(String[] args)throws IOException{

        List<String> output = (banker("gift1.in"));

        PrintWriter outFile = new PrintWriter(new BufferedWriter(new FileWriter("gift1.out")));

        for (String i : output) {
            outFile.println(i); 
            System.out.println(i);
        }
        outFile.close();

    }
}

File: programming/Puzzles/usaco/gift1.kt
import kotlin.*
import java.io.File
data class Block(
    val money : Int,
    val given : ArrayList<String>,
    val giver : String 
)


  
fun bank(name : String){
    var input : ArrayList<String> = ArrayList(File(name).readLines())
    val length = input[0].toInt()

    val people : ArrayList<String> = ArrayList(input.subList(1, length+1))
    var blocks = ArrayList<Block>()
    interpret(blocks, ArrayList(input.subList(length+1, input.size)))

    var bank = mutableMapOf<String, Int>()
    for(p in people) bank[p] = 0

    for(block in blocks){
        val given = block.given
        val money = block.money
        if(money != 0) { 

            val n = money / given.size
            val left = (money%n) - money

            for(g in given) {
                bank[g] = bank.get(g)!!.plus(n) 
            }
            bank[block.giver] = bank.get(block.giver)!!.plus(left)
        }


    }

    print(bank)
}


fun interpret(blocks : ArrayList<Block>, input : ArrayList<String>)  {
    if(input.size > 2) {
        val giver : String = input[0]
        val (money, n) = (input[1].split(" "))
        input.removeAt(0) ; input.removeAt(0)

        var given : ArrayList<String> = ArrayList<String>()
        for(i in 0..(n.toInt())-1) {
            given.add(input[0])
            input.removeAt(0)
        }
        val b : Block = Block(money.toInt(), given, giver)
        blocks.add(b)

        interpret(blocks, input)
    }
} 



fun main(){
    bank("input.txt")
    
}

